### PR TITLE
Show correct client request id on debug message

### PR DESCRIPTION
**Pull request by dingmeng-xue**
https://github.com/Azure/azure-powershell/issues/13745

The value of x-ms-client-request-id showing on debug log is different from the real value on request. The root cause is due the the sequence of request id generation, request updating in sdk, and request updating by Azure PowerShell. The order is below:

1. client request id is generated (ID1) in the constructor of cmdlet
2. ServiceClientTracingInterceptor is registered into service client
3. CmdletInfoHandler is registered into http pipeline of service client. ID1 is passed into CmdletInfoHandler.
2. Service client generates request id (ID2) because request doesn't contain x-ms-client-request-id
3. Service client calls ServiceClientTracingInterceptor to print out http request
4. Service client calls sendAsync() and start HTTP pipeline
5. CmdletInfoHandler sets ID1 into request

So, ID2 is printed out on console as debug message but ID1 is the real value used by http request.

Solution is to set ID1 into ServiceClientTracingInterceptor . ID1 is set into http request before being printed out.

Additionally, couple changes are included together
1. RecordingTracingInterceptor is used by data plane cmdlet. Request id is set too.
2. Add extensions to HttpRequestMessage
3. Add null reference check in GetLog() of GeneralUtilities

### CODE REVIEW

File: src/ResourceManager/Version2016_09_01/ServiceClientTracingInterceptor.cs [hunk 1]
```diff
@@ -62,8 +65,17 @@ public void ReceiveResponse(string invocationId, System.Net.Http.HttpResponseMes
...
+            // Update request here to ensure its value consistent with real result.
+            if (request != null && clientRequestId != null)
+            {
+                if (request.Headers.Contains(ApiConstants.HeaderNameClientRequestId))
+                {
+                    request.Headers.Remove(ApiConstants.HeaderNameClientRequestId);
+                }
+                request.Headers.TryAddWithoutValidation(ApiConstants.HeaderNameClientRequestId, clientRequestId);
+            }
+            MessageQueue.CheckAndEnqueue(GeneralUtilities.GetLog(request, Matchers));
```
Comment (erich-wang) [inline src/ResourceManager/Version2016_09_01/ServiceClientTracingInterceptor.cs:78]: request might be null here.

File: src/Common/CmdletInfoHandler.cs [hunk 2]
```diff
@@ -64,11 +64,11 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
...
             {
-                if (request.Headers.Contains("x-ms-client-request-id"))
+                if (request.Headers.Contains(ApiConstants.HeaderNameClientRequestId))
                 {
-                    request.Headers.Remove("x-ms-client-request-id");
+                    request.Headers.Remove(ApiConstants.HeaderNameClientRequestId);
                 }
-                request.Headers.TryAddWithoutValidation("x-ms-client-request-id", ClientRequestId);
+                request.Headers.TryAddWithoutValidation(ApiConstants.HeaderNameClientRequestId, ClientRequestId);
             }
```
Comment (erich-wang) [inline src/Common/CmdletInfoHandler.cs:72]: nit: It seems these code are similar in three places, how about move to helper method?

Comment (dingmeng-xue) [reply src/Common/CmdletInfoHandler.cs:72]: It's current design problem. Ideally, this logic should only show in one place. but that needs cmdlet info handler and trancing inceptors to reference each other.  So, just copy code 3 times.

I don't hope to introduce additional helper class for this problem.

File: src/ResourceManager/Version2016_09_01/ServiceClientTracingInterceptor.cs (hunk 1 above)
Comment (dingmeng-xue) [reply src/ResourceManager/Version2016_09_01/ServiceClientTracingInterceptor.cs:75]: I added check in GetLog method.

File: src/Common/CmdletInfoHandler.cs (hunk 2 above)
Comment (dingmeng-xue) [reply src/Common/CmdletInfoHandler.cs]: Add extensions

**Review by erich-wang — APPROVED** [review]

**Pull request merged**